### ISSUE TITLE
Encountering NaN with HYPRE

### ISSUE THREAD

**Issue by wcdawn** [mfem/mfem#2369]
I have a problem with several (24) linear systems of equations, each of which is a discretization of several (80) coupled PDEs. In an iterative/convergence procedure, I'm solving a selected linear system of equations while updating the RHS until the solution converges.

I'm using the `HyprePCG` solver with `HypreBoomerAMG` preconditioner. I have called
`HYPRE_BoomerAMGSetNumFunctions(amg, 80)` to improve my performance.
Since I have several linear systems, I have created a `HyprePCG` object and `HypreBoomerAMG` object for each linear system so I have vectors of these objects.

At some point during the calculation, HYPRE returns a NaN. It's well into the simulation actually. It's after converging several of the linear systems of equations and calling `HyprePCG::Mult()` many times (more than 100). The convergence behavior looks good and then just dies.

```
=============================================
Setup phase times:
=============================================
PCG Setup:
  wall clock time = 1.710000 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 1.710000 seconds
  cpu MFLOPS      = 0.000000

<C*b,b>: 5.046439e+07


Iters       ||r||_C     conv.rate  ||r||_C/||b||_C
-----    ------------    ---------  ------------
    1    4.946099e+03    0.148756    6.962581e-01
    2    1.647778e+03    0.333147    2.319562e-01
    3    8.080995e+02    0.490418    1.137555e-01
    4    4.576440e+02    0.566321    6.442216e-02
    5    3.055430e+02    0.667643    4.301104e-02
    6    2.193602e+02    0.717936    3.087916e-02
    7    1.638848e+02    0.747103    2.306992e-02
    8    1.268698e+02    0.774140    1.785935e-02
    9    9.994700e+01    0.787792    1.406945e-02
   10    8.080912e+01    0.808520    1.137543e-02
   11    6.622022e+01    0.819465    9.321764e-03
   12    5.491561e+01    0.829288    7.730424e-03
   13    4.612231e+01    0.839876    6.492598e-03
   14    3.899638e+01    0.845499    5.489488e-03
   15    3.338147e+01    0.856015    4.699082e-03
   16    2.868384e+01    0.859274    4.037800e-03
   17    2.482006e+01    0.865298    3.493899e-03
   18    2.157021e+01    0.869064    3.036421e-03
   19    1.878100e+01    0.870692    2.643786e-03
   20    1.635449e+01    0.870800    2.302208e-03
   21    1.431730e+01    0.875435    2.015435e-03
   22    1.255098e+01    0.876630    1.766791e-03
   23    1.105963e+01    0.881177    1.556855e-03
   24    9.749975e+00    0.881582    1.372496e-03
   25    8.597781e+00    0.881826    1.210302e-03
   26    7.575752e+00    0.881129    1.066432e-03
   27    6.629741e+00    0.875126    9.332630e-04
   28    5.738538e+00    0.865575    8.078092e-04
   29    4.879250e+00    0.850260    6.868479e-04
   30    4.026255e+00    0.825179    5.667725e-04
   31    3.180570e+00    0.789957    4.477261e-04
   32    2.328470e+00    0.732092    3.277767e-04
   33    1.452939e+00    0.623989    2.045290e-04
   34    4.816644e-01    0.331510    6.780349e-05
   35    nan    nan    nan


=============================================
Solve phase times:
=============================================
PCG Solve:
  wall clock time = 1.730000 seconds
  wall MFLOPS     = 0.000000
  cpu clock time  = 1.730000 seconds
  cpu MFLOPS      = 0.000000

PCG Iterations = 35
Final PCG Relative Residual Norm = nan
```

This situation raises a few issues:
1. What is causing this? I've verified my matrices are SPD by printing them and performing a Cholesky factorization. Why does the iterative procedure abruptly fail?
2. Does HYPRE/MFEM support creating an array of these solver objects? Would I be better to create a single object and call `SetPreconditioner` and `SetOperator` all the time?
3. It seems like none of the calls to HYPRE routines in MFEM actually check the return value. I've been running these on a cluster so even when a nan is encountered, the code continues to grind for another several hours unless I check and kill it. Is there a better way to perform error checking?

Thanks!

**Comment by tzanio** [comment]
Regarding 1 -- the reason for NaNs usually is that the preconditioner is not SPD. That should not be the case if the input matrix is SPD, so this sounds like a bug. Can you attach the matrix and the AMG parameters you are using so we can try to reproduce?

**Comment by tzanio** [comment]
Regarding 3 -- you usually get a segfault in situations like this, or at least hypre will raise an error. I am not sure why it is not happening in your case, it seems like you are using the hypre PCG... In MFEM, when you define a `HypreSolver` you can specify how to handle hypre errors, see https://github.com/mfem/mfem/blob/master/linalg/hypre.hpp#L840-L849.

**Comment by wcdawn** [comment]
I'm not sure about uploading the matrix. I put in in a compressed tarball and it is 17 GB. I can upload it somewhere (probably Google Drive) if someone really wants to look into it. I understand that this is not particularly helpful for debugging.

Regarding error handling, I tried switching the `HyprePCG` to `CGSolver` since they should perform the same function. Using `CGSolver`, the code crashes very quickly. It looks like there is an error in the setup of the `HypreBoomerAMG`. I'm not sure why this error is not handled when using `HyprePCG`.

```
Verification failed: (!err) is false:
 --> Error during setup! Error code: 256
 ... in function: virtual void mfem::HypreSolver::Mult(const mfem::HypreParVector&, mfem::HypreParVector&) const
 ... in file: linalg/hypre.cpp:2768

```

**Comment by tzanio** [comment]
Ok, this is progress. Can you rebuild hypre in debug mode with the option `--with-print-errors` and let us know what is the ourput?

**Comment by wcdawn** [comment]
I rebuilt HYPRE with debug options and continued using the MFEM `CGSolver`. It doesn't simply crash any more but still returns NaN at the same point. The `CGSolver` prints that the preconditioner is not SPD.
```
PCG: The preconditioner is not positive definite. (Br, r) = -0.141307
```
Obviously, things just get out of control after that.

To reiterate, the preconditioner has been called 25+ times at this point without changing the LHS matrix. The RHS vector is being reassembled every time. I'm still very confused.

Right now, I'm just using `ParBilinearForm::ParallelAssemble()` and `ParLinearForm::ParallelAssemble()` to return `HypreParMatrix*` and `HypreParVector*` respectively. I'm going to look into `ParBilinearForm::FormLinearSystem()` but I wouldn't expect anything different.

**Comment by tzanio** [comment]
OK, this message means that there is a vector `r` for which the inner product `(Br,r)` with the preconditioner `B` is negative, so `B` indeed is not SPD and PCG fails.

You can always switch to GMRES which will run, but this is a symptom of an issue in your input matrix `A`, or maybe in the preconditioner `B` and you should try to address it.

What can you tell us about `A`? Is it a standard diffusion matrix? Can it have zero rows? Do you expect it to be SPD?

What parameters do you use for `B`? Is it just the defaults in `HypreBoomerAMG`? What version of MFEM and hypre is this with?

**Comment by tzanio** [comment]
One way to test if `A` is SPD is to generate random vectors `r` and check that `(Ar,r)` is always strictly positive. Sometimes this useful, but not always, it may e.g. be difficult to detect a matrix that has one zero row/column.

**Comment by wcdawn** [comment]
It seems to be able to work with GMRES as we would expect but it also takes a significant runtime and memory hit as one would expect. The GMRES solver made it farther than PCG but it didn't finish due to wall time. I expect I could push it further if need be.

The `A` matrix has a `MassIntegrator`, `DiffusionIntegrator`, as well as a `DGTraceIntegrator` as a `BdrFaceIntegrator` to apply an "incoming" boundary condition. (I'm solving in H1 space.) I expect the operator to be SPD. These difficulties only arise for very large problems so I wonder if the operator is *nearly* SPD but only noticeable for very large systems.

The `B` preconditioner is `HypreBoomerAMG` with the defaults specified in MFEM. The only exception is I specify
```c
HYPRE_BoomerAMGSetNumFunctions(*amg, 80);
```
as this is a system of PDEs. I'm using the `master` version of MFEM and HYPRE.
Specifically, MFEM `v4.2-3343-gc703d7a` and HYRPE `v2.20.0-62-g521ef12`.

Does `HypreBoomerAMG` possibly change with every evaluation? I would think it would perform the same operation every time it was called. What is most confusing about this whole procedure is that the PCG solver works successfully many times but seems to fail randomly. I understand that this could finally be a vector that proves `B` (and/or `A`) is not SPD but I'm curious if the preconditioner is changing as well.

I'm going to try using a bunch of random vectors to see if `A` is SPD but I understand this is not necessarily going to work.

**Comment by wcdawn** [comment]
I think I found the reason why I was encountering NaN.
In fact, the `A` matrix was not SPD despite my expectations. However, the problem was that the mesh was composed of wedge/prism elements. I have yet to find a definitive proof but have encountered similar problems in smaller calculations where the `DiffusionIntegrator` in certain "malformed" meshes with wedges will not be positive-definite.

I was able to remesh the problem using tetrahedra and everything seems to be working. I'd prefer to use prisms as they more naturally describe the geometry that I'm working with so I'm interested in any tips. But that may be a separate issue.

Thanks!

**Comment by tzanio** [comment]
> I was able to remesh the problem using tetrahedra and everything seems to be working. I'd prefer to use prisms as they more naturally describe the geometry that I'm working with so I'm interested in any tips. But that may be a separate issue.

I am glad you fixed the issue @wcdawn. Thanks pushing on it!

> In fact, the A matrix was not SPD despite my expectations. However, the problem was that the mesh was composed of wedge/prism elements. I have yet to find a definitive proof but have encountered similar problems in smaller calculations where the DiffusionIntegrator in certain "malformed" meshes with wedges will not be positive-definite.

If you can create a simple reproducer we should investigate what's going on. There is no reason why the should lead to non-SPD operator. (Maybe the diffusion quadrature rule should be adjusted for prisms?!)

@mlstowell have you encountered such non-SPD issue with meshes containing prismatic elements?

**Comment by tzanio** [comment]
This should be resolved in `master` as of https://github.com/GLVis/glvis/commit/b513773a803dd99117c4ef8348465f3ae5674d86 (Jul 8th, 2021)

**Comment by wcdawn** [comment]
I have encountered this problem with other FEM codes as well so prisms may just be challenging in general. I'm not sure how to go about checking this with glvis. The plots look fine in glvis.

[Here](https://drive.google.com/file/d/1a8KrOrBoPhG1nPDe_xb44FtTxO3KWPHn/view?usp=sharing) is an example mesh. I have refined versions of this mesh or much more complex meshes that I can provide.
So far, all of the meshes that have exhibited this problem have been:
* prism elements
* cylindrical boundary/geometry
* extruded mesh

To observe that the operator is not positive definite, you can run `ex1p` using the provided mesh.

I've been struggling with this problem for a very long time so I look forward to hearing what you all find!

**Comment by tzanio** [comment]
Hi @wcdawn,

According to [mesh explorer](https://github.com/mfem/mfem/blob/master/miniapps/meshing/mesh-explorer.cpp) all the prisms in the mesh are inverted. There can be more, but this will definitely explain the issue in Example 1. 

Can you tell us how the mesh was generated? Do you have the 2D cross-section as a separate mesh file?

Tzanio

**Comment by wcdawn** [comment]
I tried using the mesh explorer and I'm not sure how to tell if an element is inverted. How can I look for myself?
I also don't understand why having inverted prisms would be a problem because I thought the method was general.

I think I may have over-simplified the problem. The cylinder I sent was a very old mesh and I'm not sure where the original data is at this point. [Here](https://drive.google.com/file/d/1PgDAoBhENAwPWVG7JUudhyDJcI6iL1Wv/view?usp=sharing) is the mesh that I have been using that is originally giving me the problems. It is large and uses a large amount of memory (>80 GB) to run `ex1p`. I generated this mesh using Gmsh and have all of the inputs. The prisms were generated using the `Recombine` command in Gmsh. Unfortunately, `ex1p` is too simple to show the error on this more complicated mesh. I'm working on putting together a MWE.

**Comment by tzanio** [comment]
> I also don't understand why having inverted prisms would be a problem because I thought the method was general.

Use the `x` option which samples the Jacobian in uniform lattice with a user provided resolution.

> I also don't understand why having inverted prisms would be a problem because I thought the method was general.

Inverted prisms lead to negative Jacobians, which lead to negative definite matrix.

> Here is the mesh that I have been using that is originally giving me the problems. It is large and uses a large amount of memory (>80 GB) to run ex1p. 

I'll try to take a look at this today...

**Comment by tzanio** [comment]
Nice mesh.
<img width="808" alt="Screen Shot 2021-07-12 at 11 50 52 AM" src="https://user-images.githubusercontent.com/1278247/125340534-6bee7d00-e307-11eb-966e-f7a78dbb04de.png">
It would appear all of the elements in it are valid:

```
--> x
subdivision factor ---> 4

bad elements = 0
min det(J)   = 0.0187922
max det(J)   = 0.930723
global ratio = 49.5272
max el ratio = 1
min kappa    = 1.1204
max kappa    = 33.1718
```

Example 1 runs fine for me on it, so a MWE will be appreciated

<img width="838" alt="Screen Shot 2021-07-12 at 11 56 22 AM" src="https://user-images.githubusercontent.com/1278247/125341256-4746d500-e308-11eb-9e31-588fc921d980.png">

**Comment by tzanio** [comment]
> > I also don't understand why having inverted prisms would be a problem because I thought the method was general.
> 
> Use the `x` option which samples the Jacobian in uniform lattice with a user provided resolution.

Another option to check mesh quality is to plot the scaled Jacobian, option `J` in Mesh Explorer. This gives a range for the element quality (specifically skewness) with 1 being a good, box-like element, and 0 being a bad invalid element. Negative values indicate inverted elements.

Here are the results for your mesh

<img width="845" alt="Screen Shot 2021-07-12 at 12 05 11 PM" src="https://user-images.githubusercontent.com/1278247/125342217-73168a80-e309-11eb-92b6-86ac6f2c825f.png">

<img width="847" alt="Screen Shot 2021-07-12 at 12 06 38 PM" src="https://user-images.githubusercontent.com/1278247/125342335-a0fbcf00-e309-11eb-8dbb-06463eee3a24.png">

**Comment by bensworth** [comment]
Depending on whether you figure out the definiteness problem, using MINRES will likely be comparable to CG and may be viable (hard to say if the preconditioner will end up positive definite). Alternatively, if you are definitely stuck with GMRES, set the k-dim to be quite small, ~<5. If the matrix is almost SPD, you shouldn't need a large kdim for good convergence. The computation of GMRES iterates is not that much more expensive than CG; almost all of your performance degradation with GMRES is likely due to memory issues (i.e., needing your full Krylov space every step and it likely being out of cache).

**Comment by wcdawn** [comment]
@bensworth this sounds promising. I've been trying to find why the matrix isn't PD but it's a real challenge. Is there a way I can call a MINRES solver from MFEM without writing my own? I'll give GMRES with a small vector dimension as well. Thanks!

**Comment by bensworth** [comment]
MFEM has a MINRES solver internally: https://mfem.github.io/doxygen/html/classmfem_1_1MINRESSolver.html

**Comment by tzanio** [comment]
@wcdawn, I am interested to follow-up on this if you have any new information.

**Comment by wcdawn** [comment]
I ran several different linear system solvers for a related 2D problem that had never had a NaN issue.
My solver requires solving a linear system thousands of times so I've also included the total number of calls to the solver.
Here are some performance results for anyone interested.

Solver | Solve Count | Wall Time [s] | Speedup (Slowdown)
----|----|----|----
`HyprePCG` | 6518 | 1695 | 0.00% (ref.)
`HypreGMRES` (kdim=5) | 6519 | 2688 | -58.58%
`CGSolver` | 6519 | 2149 | -26.78%
`GMRESSolver` (kdim=5) | 6508 | 3048 | -79.82%
`MINRESSolver` | 6519 | 2080 | -22.71%
`BiCGSTABSolver` | 7227 | 3007 | -77.40%

For my problems, `MINRESSolver` seems comparable to `CGSolver` and they can probably be sped up using some better solver options to be comparable with `HyprePCG`.

I tried to do the same analysis using the full 3D mesh that I had sent earlier. All of the solvers failed. Even GMRES failed but I suspect the vector dimension was too small.

I tried the 3D problem without a preconditioner and using `HyprePCG` (I had been using `mfem::HypreBoomerAMG`).
The unpreconditioned problem made it farther (more successful calls to the linear system solver) than the problem with a preconditioner but it failed to finish due to wall time as it was so much slower. I'm going to look into using an alternate preconditioner (maybe `GSSmoother`?) but I'm not sure how it will do in parallel.

**Comment by bensworth** [comment]
Can you look at iterations to convergence for HyprePCG and CGSolver? Interesting difference in walltime. Similar difference in HypreGMRES and GMRESSolver. Would be strange if the iterations were different. If those are the same, my guess is hypre is built with its own LAPACK, which could speed up the inner dense Krylov iterations.

**Comment by wcdawn** [comment]
The iteration counts were pretty different with the MFEM solvers being much larger. I think the solvers are using different measures for convergence criteria. It looks like the MFEM solvers are using the `(B r, r)` or `||r||_B` norm whereas I think HYPRE is using `||r||_2`. I could select some criteria to make them run the same number of iterations or set the tolerances to zero and run a fixed number of iterations if we really needed to compare.

For now, I'm less interested in solver performance as I am in finding a solver+preconditioner that will actually work for these equations.

**Comment by victorapm** [comment]
Hi @wcdawn,

Clarifying one of your doubts:

>Does HypreBoomerAMG possibly change with every evaluation?

BoomerAMG would change only if you were calling `HYPRE_BoomerAMGSetup`again with a different matrix or set of input parameters, but not during PCG iterations.

Could you run your problematic test case again with `pcg->SetPrintLevel(2);`? This prints some statistics regarding BoomerAMG setup that can be helpful.

Thanks!
Victor

**Comment by wcdawn** [comment]
@victorapm Thanks for the info! I'm definitely not calling the setup function more than once. Only the RHS is changing.
The output is pretty large so I've attached in in a text file.
[hypre.txt](https://github.com/mfem/mfem/files/6926494/hypre.txt)

**Comment by victorapm** [comment]
Thanks, @wcdawn, I don't see BoomerAMG setup statistics in this file. Could you try again with the following options:
```
pcg->SetPrintLevel(2);
amg->SetPrintLevel(1);
```

I'm attaching the output of `ex2p` that I get with these options so that you have an idea of what to expect.
[ex2p_out.txt](https://github.com/mfem/mfem/files/6926608/ex2p_out.txt)

**Comment by wcdawn** [comment]
@victorapm This should be the correct data that you're looking for.
[hypre.txt](https://github.com/mfem/mfem/files/6926773/hypre.txt)

**Comment by victorapm** [comment]
Hi @wcdawn, since you are reusing the preconditioner for several system solves, it's worth using a set of config options that improves its quality. In your case, you could try:
```
amg->SetAggressiveCoarsening(0);
amg->SetStrengthThresh(0.5);
```

This does not necessarily solve the problem with indefiniteness as it is likely coming from `A` , but will probably give you better solve times. Besides this, you can try also using a bigger `kdim` for GMRES (maybe 10 or 20). Let me know what you find!

Thanks!
